### PR TITLE
Python api of placement group

### PR DESCRIPTION
**Pull request by Alisahhh**
## Why are these changes needed?

This PR is based on https://github.com/ray-project/ray/pull/9120/commits
This PR give a python API.
``` bash
cd python
pip install -e . --verbose    
touch test.py
```
test.py
``` python
import ray

ray.init(include_dashboard=False)

@ray.remote(num_cpus=1)
class Counter(object):
    def __init__(self):
        self.n = 0

    def increment(self):
        self.n += 5

    def read(self):
        return self.n

placement_group_id = ray.experimental.placement_group( name = "name", strategy = "PACK",bundles = [{"CPU":4}])
print(placement_group_id)

counters = [Counter.options(placement_group_id = placement_group_id, bundle_index = 0).remote() for i in range(4)]
[c.increment.remote() for c in counters]
futures = [c.read.remote() for c in counters]
print(ray.get(futures))
```

TODO:
- Check if the bundle index is correct. (Maybe in GCS)
- Read the data from redis when the gcs dump.

BTW, This PR fix a bug in resource.

## Checks

- [ *] I've run `scripts/format.sh` to lint the changes in this PR.
- [ *] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -907,14 +913,19 @@ cdef class CoreWorker:
                      c_bool is_detached,
                      c_string name,
                      c_bool is_asyncio,
-                     c_string extension_data):
+                     PlacementGroupID placement_group_id,
+                     int64_t bundle_index,
```
Comment (ericl) [inline python/ray/_raylet.pyx:917]:
```suggestion
                     int64_t placement_group_bundle_index,
```

File: python/ray/_raylet.pyx [hunk 2]
```diff
@@ -929,12 +940,40 @@ cdef class CoreWorker:
...
+                            c_string name,
+                            c_vector[unordered_map[c_string, double]] bundles,
+                            c_string strategy):
+        cdef:
+            CPlacementGroupID c_placement_group_id
+            CPlacementStrategy c_strategy
+        with self.profile_event(b"submit_task"):
+            if strategy == b"PACK":
+                c_strategy = PLACEMENT_STRATEGY_PACK
+            else:
```
Comment (ericl) [inline python/ray/_raylet.pyx:961]: Can you check explicitly for SPREAD as well and raise an error if not that or pack?

File: python/ray/_raylet.pyx [hunk 3]
```diff
@@ -929,12 +940,40 @@ cdef class CoreWorker:
...
 
+    def create_placement_group(
+                            self,
+                            c_string name,
+                            c_vector[unordered_map[c_string, double]] bundles,
+                            c_string strategy):
+        cdef:
+            CPlacementGroupID c_placement_group_id
+            CPlacementStrategy c_strategy
+        with self.profile_event(b"submit_task"):
```
Comment (ericl) [inline python/ray/_raylet.pyx:958]: Should remove this profile event since it's the wrong type.

File: python/ray/actor.py [hunk 4]
```diff
@@ -411,7 +411,9 @@ def _remote(self,
                 max_restarts=None,
                 max_task_retries=None,
                 name=None,
-                detached=False):
+                detached=False,
+                placement_group_id=None,
+                bundle_index=None):
```
Comment (ericl) [inline python/ray/actor.py:416]:
```suggestion
                placement_group_bundle_index=None):
```

File: python/ray/actor.py [hunk 5]
```diff
@@ -411,7 +411,9 @@ def _remote(self,
...
         """Create an actor.
```
Comment (ericl) [inline python/ray/actor.py:417]: Can you document these args? In particular, does None mean any index is allowed?

File: python/ray/actor.py [hunk 6]
```diff
@@ -498,6 +499,11 @@ def _remote(self,
         else:
             detached = False
 
+        if placement_group_id is not None and bundle_index is None:
+            raise ValueError("The placement_group_id is set."
+                             "But the bundle_index is not set.")
+            pass
```
Comment (ericl) [inline python/ray/actor.py:505]:
```suggestion
```

File: python/ray/actor.py [hunk 7]
```diff
@@ -498,6 +499,11 @@ def _remote(self,
         else:
             detached = False
 
+        if placement_group_id is not None and bundle_index is None:
+            raise ValueError("The placement_group_id is set."
+                             "But the bundle_index is not set.")
```
Comment (ericl) [inline python/ray/actor.py:504]: Can we allow None to mean any bundle?

File: python/ray/experimental/placement_group.py [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import ray
+
+
+def placement_group(name, strategy, bundles):
```
Comment (ericl) [inline python/ray/experimental/placement_group.py:4]:
```suggestion
def placement_group(bundles: List[Dict[str, float]], strategy: str = "PACK", name: str = None):
```

File: python/ray/experimental/placement_group.py [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+    worker = ray.worker.global_worker
```
Comment (ericl) [inline python/ray/experimental/placement_group.py:5]: Please add pydoc for this method.

File: python/ray/tests/test_placement_group.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
+import pytest
```
Comment (ericl) [inline python/ray/tests/test_placement_group.py:1]: You'll need to add an entry in BUILD.bazel for this test to be run.

File: python/ray/actor.py [hunk 11]
```diff
@@ -411,7 +411,9 @@ def _remote(self,
                 max_restarts=None,
```
Comment (ericl) [inline python/ray/actor.py:411]: We can support tasks in a follow-up pr.

**Review by ericl — CHANGES_REQUESTED** [review]

**Comment by ericl** [comment]
Looks good, I left some comments.

File: python/ray/actor.py (hunk 7 above)
Comment (wumuzi520) [reply python/ray/actor.py:509]: You mean random a boundle_index if the it is None?

Comment (raulchen) [reply python/ray/actor.py:509]: @wumuzi520, I discussed with @ericl before, supporting any bundle is a valid use case.

File: python/ray/tests/test_placement_group.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    for _ in range(num_nodes):
+        cluster.add_node(num_cpus=4)
+    ray.init(address=cluster.address)
+
+    placement_group_id = ray.experimental.placement_group(
+        name="name", strategy="PACK", bundles=[{
+            "CPU": 2
+        }, {
+            "CPU": 2
+        }])
```
Comment (raulchen) [inline python/ray/tests/test_placement_group.py:39]: nit, if you add `,` after `}`, the format will look nicer.

File: python/ray/experimental/placement_group.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
+import ray
+
+
+def placement_group(bundles: List[Dict[str, float]], strategy: str = "PACK",
```
Comment (raulchen) [inline python/ray/experimental/placement_group.py:4]: Define an enum for strategy.

Comment (raulchen) [reply python/ray/experimental/placement_group.py:4]: We can define a `PlacementGroup` type in this file. The advantages are:
1) avoid exposing placement group id (which is an implementation detail) to users. 
2) Document what a placement group is there.

File: python/ray/actor.py (hunk 7 above)
Comment (raulchen) [reply python/ray/actor.py:509]: I'm not sure whether the backend already supports this though.

File: python/ray/actor.py [hunk 14]
```diff
@@ -436,6 +438,11 @@ def _remote(self,
                 guaranteed when max_concurrency > 1.
             name: The globally unique name for the actor.
             detached: DEPRECATED.
+            placement_group_id: the placement group this actor belongs to, 
+                if none means no placement group belongs to.
```
Comment (ericl) [inline python/ray/actor.py:442]:
```suggestion
                or None if it doesn't belong to any group.
```

File: python/ray/actor.py [hunk 15]
```diff
@@ -436,6 +438,11 @@ def _remote(self,
...
+            placement_group_bundle_index: the index of the bundle
+                if the actor benlongs to a placment group.
```
Comment (ericl) [inline python/ray/actor.py:444]:
```suggestion
                if the actor belongs to a placement group. 
```

File: python/ray/actor.py [hunk 16]
```diff
@@ -436,6 +438,11 @@ def _remote(self,
                 guaranteed when max_concurrency > 1.
             name: The globally unique name for the actor.
             detached: DEPRECATED.
+            placement_group_id: the placement group this actor belongs to, 
+                if none means no placement group belongs to.
+            placement_group_bundle_index: the index of the bundle
+                if the actor benlongs to a placment group. 
+                If it's none, it's means don't belong to a placement group.
```
Comment (ericl) [inline python/ray/actor.py:445]:
```suggestion
```

File: python/ray/actor.py (hunk 7 above)
Comment (ericl) [reply python/ray/actor.py:509]: Ok, maybe we can add it later.

File: python/ray/experimental/placement_group.py (hunk 13 above)
Comment (ericl) [reply python/ray/experimental/placement_group.py:5]: It's more Pythonic to use a string choice here I think.

File: python/ray/experimental/placement_group.py [hunk 17]
```diff
@@ -0,0 +1,26 @@
+import ray
+from typing import (List, Dict)
+
+
+def placement_group(bundles: List[Dict[str, float]], strategy: str = "PACK",
+                    name: str = None):
+    """
+    placement_group.
```
Comment (ericl) [inline python/ray/experimental/placement_group.py:8]:
```suggestion
    Create a placement group.
```

(please also remove the newline, so it matches the style of other docstrings)

File: python/ray/experimental/placement_group.py [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+
+    This method is the api to create placement group.
+
+    Args:
+    bundles: A list of bundles which represent the resources needed.
```
Comment (ericl) [inline python/ray/experimental/placement_group.py:13]: Please indent these 4 spaces under args (see other docstringS)

**Review by ericl — APPROVED** [review]
Looks good, feel free to merge once typos are addressed.

File: python/ray/tests/test_placement_group.py [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+
+    cluster = ray_start_cluster
+    num_nodes = 2
+    for _ in range(num_nodes):
+        cluster.add_node(num_cpus=4)
+    ray.init(address=cluster.address)
+
+    placement_group_id = ray.experimental.placement_group(
+        name="name", strategy="PACK", bundles=[{
+            "CPU": 2
```
Comment (rkooo567) [inline python/ray/tests/test_placement_group.py:36]: Is this going to be custom resources in this case? Or is it translated to num_cpus resources?

**Review by rkooo567 — APPROVED** [review]
The PR LGTM! I have some questions for testing strategy. I think we are missing tests scenarios such as 

1. Resources that exceed boarder line. E.g, PACK strategy with 3 and 2 cpus requirements for 4 cpus node. 
2. Duplicated bundle index.
3. Infeasible bundles? (3 X 3 cpus) in 8 cpus cluster. 
(and I believe we can find many more!)
Do we have any plan to add additional tests later?

**Comment by ericl** [comment]
Tests are flaky on master; merging!

**Pull request merged**